Mathlib/Algebra/Ring/SumsOfSquares.lean

```diff
@@ -0,0 +1,97 @@
+/-
+Copyright (c) 2024 Florent Schaffhauser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Florent Schaffhauser
+-/
+
+import Mathlib.Algebra.Ring.Defs
+import Mathlib.Algebra.Order.Ring.Defs
+
+/-!
+# Sums of squares
+
+We introduce the type sums of squares in a semiring `R` as the subtype of `R` whose terms are
+sums of squares in `R`. The associated predicate `isSumSq : R → Prop` is defined inductively:
+`0 : R` is a sum of squares and if `S` is a sum of squares, then for all `a : R`, `a ^ 2 + S` is a
+sum of squares in `R`.
+
+## Main declaration
+
+- The type `SumSqIn R`: it is the type of sums of squares in `R`, where `R` is a semiring.
+
+## Auxiliary declarations
+
+- The predicate `isSumSq : R → Prop`: the predicate used to define the type `SumSqIn R` as a subtype
+of `R`.
+
+## Theorems
+
+- `isSumSq.add`: if `S₁` and `S₂` are sums of squares in a semiring `R`, then so is `S₁ + S₂`.
+- `SumSq.nonneg`: sums of squares are non-negative.
+
+## Instances
+
+- An `Add` instance on the type of sums of squares in a semiring `R`.
+
+-/
+
+/--
+In a semiring `R`, the property of being a sum of squares is an inductive predicate: `0 : R` is a
+sum of squares and if `S` is a sum of squares, then for all `a : R`, `a ^ 2 + S` is a sum of squares
+in `R`.
+-/
+
+@[mk_iff]
+inductive isSumSq {R : Type*} [Add R] [Mul R] [Zero R] : R → Prop
+  | zero                              : isSumSq 0
+  | sq_add (x S : R) (pS : isSumSq S) : isSumSq (x * x + S)
+
+/--
+The type of sums of squares in a semiring `R` is the subtype of `R` defined by the predicate
+`isSumSq : R → Prop`. It  can be defined as a structure whose terms are dependent pairs `⟨x, hx⟩`
+where `x : R` and `hx` is a proof of  the proposition `isSumSq x`.
+-/
+
+structure SumSqIn (R : Type*) [Add R] [Mul R] [Zero R] where
```

Comment:
I am sorry, but this definition is plainly wrong. By this I mean that what you are redefining essentially a subtype, but ignoring the library, and this is a recipe for suffering later. If you want to work only with `isSumSq` that is OK, but if you want something like `SumSqIn` (mathematically the subset of elements that are sum of squares) then `AddSubmoniid` is absolutely the way to go (or some other `...Sub` class if you want to think about minimal assumptions to have various properties, but this is not very important). I perfectly understand that you don't want to focus in this, but I think this is the way to spend *less* time with trivial properties of sum of squares.

Reply:
OK, I'll change it then ☺️ But I don't think it is necessary to call it wrong. It was just a design choice.

Reply:
Yeah, sorry for that. I mean that I think it will make things more complicated later. Of course it is not mathematically wrong.

Reply:
Right, and you have a point, too, when you say that it is more consistent with Mathlib this way, which is of course very important since this is a PR to Mathlib 😅 

I will retake your suggestion


```lean
def AddSubmonoid.isSumSq (R : Type*) [Ring R] : AddSubmonoid R where
  carrier := {r : R | _root_.isSumSq r}
  zero_mem' := isSumSq.zero
  add_mem' := sorry
```

and start from there.

Reply:
Using what you already have that `sorry` is done. Then you can start working on semireal field, and adding to this file only the results you need.

Reply:
Please ignore the question if you don't have time: what does `_root_` mean in the definition of the `carrier`?

Reply:
In general if you have two declaration `bar` and `foo.bar` and you have `open foo` then sometimes Lean is confused if you write `bar` (as it doesn't understand if you want `bar` or `foo.bar`). Writing  `_root_.bar` means "I want the declaration called exactly `bar`, ignoring what I've opened.

Here what's happening is a little subtle: since we are defining something called `AddSubmonoid.isSumSq` Lean automatically opens the namespace `AddSubmonoid` inside the definition (this can be confusing sometimes, but it is usually useful). So now `isSumSq` can be `_root_.isSumSq` or `AddSubmonoid.isSumSq` and Lean is confused (not that we do inductive definitions all the time, so we really want to use a definition inside itself). If you call the submonoid `AddSubmonoid.isSumSqIn` or whatever the `_root_.` shouldn't be needed.

Reply:
Ok (though I did not fully get the very last sentence). And we absolutely do want to add the `AddSubmonoid` prefix, right? You do not want to see the following, for instance:

```lean
def SumSqIn (R : Type*) [AddMonoid R] [Mul R] : AddSubmonoid R where
  carrier   := {S : R | isSumSq S}
  zero_mem' := isSumSq.zero
  add_mem'  := isSumSq.add
```

Reply:
The name is much less important, I am not against this.